src/protocols/bdx/AsyncTransferFacilitator.h

```diff
@@ -0,0 +1,136 @@
+#include <protocols/bdx/BdxTransferSession.h>
+#include <transport/raw/MessageHeader.h>
+
+#pragma once
+
+namespace chip {
+namespace bdx {
+
+/**
+ * An abstract class with methods for handling BDX messages received from an ExchangeContext. Once a message is received, this
+ * class passes the message to the TransferSession to process the received messages and calls ProcessOutputEvents to get the next
+ * output events generated by the TransferSession state machine until it receives an output event of type
+ * TransferSession::OutputEventType::kNone. For each output event it receives, it either sends a message accross the exchange if the
+ * TransferSession generated an event of type TransferSession::OutputEventType::kMsgToSend or calls the HandleTransferSessionOutput
+ * method to notify the subclass of the output event which has the business logic of handling the BDX message and calling the
+ * TransferSession API to proceed with the BDX transfer. For e.g., If this class received a ReceiveInit message and passed it to the
+ * HandleTransferSessionOutput method, the subclass would call the AcceptTransfer on the TransferSession object to negotiate params
+ * and prepare a ReceiveAccept message.
+ *
+ * This class does not define any methods for beginning a transfer or initializing the underlying TransferSession object.
+ * See AsyncResponder for a class that does.
+ * TODO: # 29334 - Add AsyncInitiator to handle the initiating side of a transfer.
+ *
+ * An AsyncTransferFacilitator is associated with a specific BDX transfer.
+ */
+class AsyncTransferFacilitator : public Messaging::ExchangeDelegate
+{
+public:
+    AsyncTransferFacilitator() : mExchange(*this) {}
+    ~AsyncTransferFacilitator() override;
+
+protected:
+    CHIP_ERROR Init(System::Layer * layer, Messaging::ExchangeContext * exchangeCtx, System::Clock::Timeout timeout);
+    CHIP_ERROR OnMessageReceived(Messaging::ExchangeContext * ec, const PayloadHeader & payloadHeader,
+                                 System::PacketBufferHandle && payload) override;
+    void OnResponseTimeout(Messaging::ExchangeContext * ec) override;
+    void OnExchangeClosing(Messaging::ExchangeContext * ec) override;
+
+    CHIP_ERROR SendMessage(const TransferSession::MessageTypeData msgTypeData, System::PacketBufferHandle & msgBuf);
+
+    static bdx::StatusCode GetBdxStatusCodeFromChipError(CHIP_ERROR err);
+
+    /**
+     * This method should be implemented to contain business-logic handling of BDX messages
+     * and other TransferSession events.
+     *
+     * @param[in] event An OutputEvent that contains the output from the TransferSession object.
+     */
+    virtual void HandleTransferSessionOutput(TransferSession::OutputEvent & event) = 0;
+
+    /**
+     * This method should be implemented to destroy the object subclassing AsyncTransferFacilitator.
+     */
+    virtual void DestroySelf() = 0;
+
+    void ProcessOutputEvents();
+
+    void CleanUp();
```

Comment:
This needs some really clear documentation on who is expected to call it and when.  When is it safe to call?  Is it safe to call synchronously from under `HandleTransferSessionOutput`?  Is it _required_ to call it at various points?

The PR comments above say:

> It's called by subclass for various scenarios including error conditions

but wouldn't error conditions call `NotifyEventHandled` with an error?  This part really needs to be very clearly defined so we can get the invariants right.